metrics/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,226 @@
+		gauges:     map[string]*Gauge{},
+		histograms: map[string]*Histogram{},
+		logger:     logger,
+	}
+}
+
+// NewCounter returns a counter. Observations are aggregated and emitted once
+// per write invocation.
+func (cw *CloudWatch) NewCounter(name string) *Counter {
+	c := NewCounter(name)
+	cw.mtx.Lock()
+	cw.counters[name] = c
+	cw.mtx.Unlock()
+	return c
+}
+
+// NewGauge returns a gauge. Observations are aggregated and emitted once per
+// write invocation.
+func (cw *CloudWatch) NewGauge(name string) *Gauge {
+	g := NewGauge(name)
+	cw.mtx.Lock()
+	cw.gauges[name] = g
+	cw.mtx.Unlock()
+	return g
+}
+
+// NewHistogram returns a histogram. Observations are aggregated and emitted as
+// per-quantile gauges, once per write invocation. 50 is a good default value
+// for buckets.
+func (cw *CloudWatch) NewHistogram(name string, buckets int) *Histogram {
+	h := NewHistogram(name, buckets)
+	cw.mtx.Lock()
+	cw.histograms[name] = h
+	cw.mtx.Unlock()
+	return h
+}
+
+// WriteLoop is a helper method that invokes Send every
+// time the passed channel fires. This method blocks until the channel is
+// closed, so clients probably want to run it in its own goroutine. For typical
+// usage, create a time.Ticker and pass its C channel to this method.
+func (cw *CloudWatch) WriteLoop(c <-chan time.Time) {
+	for range c {
+		if err := cw.Send(); err != nil {
+			cw.logger.Log("during", "Send", "err", err)
+		}
+	}
+}
+
+// Send will fire an api request to CloudWatch with the latest stats for
+// all metrics.
+func (cw *CloudWatch) Send() error {
+	cw.mtx.RLock()
+	defer cw.mtx.RUnlock()
+	now := time.Now()
+
+	datums := []*cloudwatch.MetricDatum{}
+
+	for name, c := range cw.counters {
+		datums = append(datums, &cloudwatch.MetricDatum{
+			MetricName: aws.String(name),
+			Dimensions: makeDimensions(c.c.LabelValues()...),
+			Value:      aws.Float64(c.c.Value()),
+			Timestamp:  aws.Time(now),
+		})
+	}
+
+	for name, g := range cw.gauges {
+		datums = append(datums, &cloudwatch.MetricDatum{
+			MetricName: aws.String(name),
+			Dimensions: makeDimensions(g.g.LabelValues()...),
+			Value:      aws.Float64(g.g.Value()),
+			Timestamp:  aws.Time(now),
+		})
+	}
+
+	for name, h := range cw.histograms {
+		for _, p := range []struct {
+			s string
+			f float64
+		}{
+			{"50", 0.50},
+			{"90", 0.90},
+			{"95", 0.95},
+			{"99", 0.99},
+		} {
+			datums = append(datums, &cloudwatch.MetricDatum{
+				MetricName: aws.String(fmt.Sprintf("%s_%s", name, p.s)),
+				Dimensions: makeDimensions(h.h.LabelValues()...),
+				Value:      aws.Float64(h.h.Quantile(p.f)),
+				Timestamp:  aws.Time(now),
+			})
+		}
+	}
+
+	_, err := cw.svc.PutMetricData(&cloudwatch.PutMetricDataInput{
+		Namespace:  aws.String(cw.namespace),
+		MetricData: datums,
+	})
+	return err
+}
+
+// Counter is a CloudWatch counter metric.
+type Counter struct {
+	c *generic.Counter
+}
+
+// NewCounter returns a new usable counter metric.
+func NewCounter(name string) *Counter {
```

Comment:
It seems the only "correct" way to construct a new {Counter, Gauge, Histogram} is via the corresponding method on the CloudWatch struct. Am I reading this correctly? If so, should these free functions be eliminated?